src/transformers/models/dbrx/modeling_dbrx.py

```diff
@@ -0,0 +1,2132 @@
+                 moe_normalize_expert_weights: Optional[float],
+                 uniform_expert_assignment: bool):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.moe_num_experts = moe_num_experts
+        self.moe_top_k = moe_top_k
+        self.moe_jitter_eps = moe_jitter_eps
+        self.moe_normalize_expert_weights = moe_normalize_expert_weights
+        self.uniform_expert_assignment = uniform_expert_assignment
+
+        self.layer = nn.Linear(self.hidden_size,
+                               self.moe_num_experts,
+                               bias=False)
+
+    def jitter(self, x: torch.Tensor) -> torch.Tensor:
+        if self.moe_jitter_eps is None:
+            raise RuntimeError('The router does not have moe_jitter_eps set.')
+        low = 1.0 - self.moe_jitter_eps
+        high = 1.0 + self.moe_jitter_eps
+        noise = torch.rand(x.size(), dtype=x.dtype, device=x.device)
+        return low + noise * (high - low)
+
+    def forward(
+            self, x: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.LongTensor]:
+        if self.training and self.moe_jitter_eps is not None:
+            x = x * self.jitter(x)
+
+        weights = self.layer(x.view(-1,
+                                    x.shape[-1])).softmax(dim=-1,
+                                                          dtype=torch.float32)
+        top_weights, top_experts = torch.topk(weights, self.moe_top_k, dim=-1)
+
+        if self.moe_normalize_expert_weights:
+            top_weights = top_weights / torch.norm(
+                top_weights,
+                p=self.moe_normalize_expert_weights,
+                dim=-1,
+                keepdim=True)
+
+        if self.uniform_expert_assignment:
+            with torch.no_grad():
+                uniform_tensor = torch.arange(
+                    0,
+                    top_experts.numel(),
+                    device=top_experts.device,
+                    dtype=top_experts.dtype) % self.moe_num_experts
+                top_experts = uniform_tensor.reshape(top_experts.shape)
+                # Note, weights and top_weights are not changed
+
+        weights = weights.to(x.dtype)
+        top_weights = top_weights.to(x.dtype)
+        return weights, top_weights, top_experts  # type: ignore
+
+
+class DbrxExpertGLU(nn.Module):
+
+    def __init__(self, hidden_size: int, ffn_hidden_size: int,
+                 moe_num_experts: int, ffn_act_fn: dict):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.ffn_hidden_size = ffn_hidden_size
+        self.moe_num_experts = moe_num_experts
+
+        self.w1 = nn.Parameter(
+            torch.empty(moe_num_experts * ffn_hidden_size, hidden_size))
+        self.v1 = nn.Parameter(
+            torch.empty(moe_num_experts * ffn_hidden_size, hidden_size))
+        self.w2 = nn.Parameter(
+            torch.empty(moe_num_experts * ffn_hidden_size, hidden_size))
```

Comment:
why not use linears?

Reply:
I think using `Linear` here would facilitate applying LoRA and similar PEFT methods to DBRX.

Reply:
The fused weights are an artifact of our training process and we'd really like to preserve it and the key names, especially as the DBRX models are already out and being integrated into inference engines and such.

At present there is also a workaround for LoRA from the community here! https://huggingface.co/databricks/dbrx-instruct/discussions/10#660566f14f41c0c7c0e54ab9

One of our teammates @cli99 has been working on a potential solution where the weights can be remapped upon `load_state_dict` into a list of `nn.Linear`. But it also requires some small edits to `transformers`. I'd like to separate that work into a second PR though, so it's easier to discuss. Also that way we can get basic DBRX functionality merged first and follow up with improvements.

Reply:
Even so, a fix to load_state_dict is not a complete solution as you would also need to introduce additional complexity when saving model checkpoints to be compatible.